File: python/inline_sql_helper/protection.py
"""Protect Python syntax while a SQL formatter rewrites a literal body."""

from __future__ import annotations

import re
from collections.abc import Callable, Sequence
from dataclasses import dataclass
from enum import StrEnum

from inline_sql_helper.detection import SqlDetection
from inline_sql_helper.literals import LiteralKind, SupportedLiteral
from inline_sql_helper.positions import SourceMap, SourceSpan


class ProtectedKind(StrEnum):
    """Kind of source text replaced by an opaque formatter marker."""

    FIELD = "field"
    ESCAPED_BRACE = "escaped_brace"
    PYTHON_ESCAPE = "python_escape"
    SQL_MARKER = "sql_marker"


class UnsafeRestore(ValueError):
    """A source-free failure to restore protected Python text."""


@dataclass(frozen=True, slots=True)
class ProtectedFragment:
    """One source span replaced by a nonce-bearing marker."""

    kind: ProtectedKind
    ordinal: int
    source_span: SourceSpan
    source_text: str
    marker: str
    required_offset: int | None


@dataclass(frozen=True, slots=True)
class ProtectionPlan:
    """The masked literal body and its exact restoration fragments."""

    nonce: str
    protected_sql: str
    fragments: tuple[ProtectedFragment, ...]


def allocate_nonce(source: str, random_bytes: Callable[[int], bytes]) -> str:
    """Allocate a cryptographic request nonce absent from *source*."""
    for _attempt in range(128):
        candidate = random_bytes(16)
        if len(candidate) != 16:
            raise UnsafeRestore("invalid protection nonce")
        nonce = candidate.hex()
        if nonce not in source:
            return nonce
    raise UnsafeRestore("unable to allocate protection nonce")


def marker_text(
    nonce: str,
    kind: ProtectedKind,
    ordinal: int,
    *,
    sql_comment: bool,
    canonical_newline: bool,
) -> str:
    """Return the stable opaque token used for one protected fragment."""
    token = f"__INLINE_SQL_{nonce}_{kind.name}_{ordinal}__"
    return ("-- " if sql_comment else "") + token + ("\n" if canonical_newline else "")


def _intersects_any(span: SourceSpan, blocked: Sequence[SourceSpan]) -> bool:
    return any(span.start < other.end and other.start < span.end for other in blocked)


def _python_escape_end(source: str, start: int, limit: int) -> int:
    """Return the complete source spelling of an escape at *start*."""
    if start + 1 >= limit or source[start] != "\\":
        raise UnsafeRestore("invalid Python escape boundary")
    next_character = source[start + 1]
    if next_character == "\r" and start + 2 < limit and source[start + 2] == "\n":
        return start + 3
    if next_character in "\r\n":
        return start + 2
    if next_character == "N" and start + 2 < limit and source[start + 2] == "{":
        close = source.find("}", start + 3, limit)
        if close < 0:
            raise UnsafeRestore("unterminated named Python escape")
        return close + 1
    widths = {"x": 2, "u": 4, "U": 8}
    if next_character in widths:
        end = start + 2 + widths[next_character]
        if end > limit:
            raise UnsafeRestore("unterminated fixed-width Python escape")
        return end
    if next_character in "01234567":
        end = start + 2
        while end < min(start + 4, limit) and source[end] in "01234567":
            end += 1
        return end
    return start + 2


@dataclass(frozen=True, slots=True)
class _FragmentSpec:
    kind: ProtectedKind
    source_span: SourceSpan
    sql_comment: bool = False
    canonical_newline: bool = False
    required_offset: int | None = None


def _discover_source_specs(
    source_map: SourceMap,
    literal: SupportedLiteral,
    detection: SqlDetection,
) -> tuple[_FragmentSpec, ...]:
    """Discover protected syntax spans without evaluating Python text."""
    content = literal.content_span
    specs: list[_FragmentSpec] = []
    if detection.marker_span is not None:
        if detection.sql_span is None:
            raise UnsafeRestore("marker detection has no SQL boundary")
        marker_span = SourceSpan(content.start, detection.sql_span.start)
        marker_source = source_map.slice(marker_span)
        specs.append(
            _FragmentSpec(
                ProtectedKind.SQL_MARKER,
                marker_span,
                sql_comment=True,
                canonical_newline=marker_source.endswith(("\r", "\n")),
                required_offset=0,
            )
        )
    specs.extend(
        _FragmentSpec(ProtectedKind.FIELD, span) for span in literal.field_spans
    )
    fixed_spans = [spec.source_span for spec in specs]

    if literal.kind not in {LiteralKind.RAW, LiteralKind.RAW_FSTRING}:
        cursor = content.start
        while cursor < content.end:
            if source_map.text[cursor] != "\\":
                cursor += 1
                continue
            end = _python_escape_end(source_map.text, cursor, content.end)
            span = SourceSpan(cursor, end)
            if not _intersects_any(span, fixed_spans):
                specs.append(_FragmentSpec(ProtectedKind.PYTHON_ESCAPE, span))
            cursor = end

    cursor = content.start
    while cursor + 1 < content.end:
        pair_span = SourceSpan(cursor, cursor + 2)
        pair = source_map.text[pair_span.start : pair_span.end]
        if pair in {"{{", "}}"}:
            overlaps = [
                spec
                for spec in specs
                if _intersects_any(pair_span, (spec.source_span,))
            ]
            if not overlaps:
                specs.append(_FragmentSpec(ProtectedKind.ESCAPED_BRACE, pair_span))
            elif all(spec.kind is ProtectedKind.PYTHON_ESCAPE for spec in overlaps):
                merged = SourceSpan(
                    min(
                        pair_span.start, *(spec.source_span.start for spec in overlaps)
                    ),
                    max(pair_span.end, *(spec.source_span.end for spec in overlaps)),
                )
                specs = [spec for spec in specs if spec not in overlaps]
                specs.append(_FragmentSpec(ProtectedKind.PYTHON_ESCAPE, merged))
            cursor += 2
            continue
        cursor += 1
    return tuple(sorted(specs, key=lambda spec: spec.source_span))


def mask_fragments(
    content: str,
    content_start: int,
    fragments: Sequence[ProtectedFragment],
) -> str:
    """Replace source fragments in order, rejecting overlap or out-of-range spans."""
    pieces: list[str] = []
    cursor = 0
    for fragment in fragments:
        start = fragment.source_span.start - content_start
        end = fragment.source_span.end - content_start
        if start < cursor or end < start or end > len(content):
            raise UnsafeRestore("protected source spans overlap")
        pieces.extend((content[cursor:start], fragment.marker))
        cursor = end
    pieces.append(content[cursor:])
    return "".join(pieces)


def build_protection_plan(
    source_map: SourceMap,
    literal: SupportedLiteral,
    detection: SqlDetection,
    nonce: str,
) -> ProtectionPlan:
    """Build an exact source mask for one detected SQL literal."""
    if (
        not detection.matched
        or detection.sql_span is None
        or re.fullmatch(r"[0-9a-f]{32}", nonce) is None
        or nonce in source_map.text
    ):
        raise UnsafeRestore("invalid protection-plan input")
    if (
        literal.content_span.start < 0
        or literal.content_span.end > len(source_map.text)
        or detection.sql_span.start < literal.content_span.start
        or detection.sql_span.end != literal.content_span.end
    ):
        raise UnsafeRestore("invalid protection-plan span")
    specs = _discover_source_specs(source_map, literal, detection)
    fragments: list[ProtectedFragment] = []
    previous_end = literal.content_span.start
    for ordinal, spec in enumerate(specs):
        span = spec.source_span
        if (
            span.start < previous_end
            or span.start < literal.content_span.start
            or span.end > literal.content_span.end
        ):
            raise UnsafeRestore("protected source spans overlap")
        fragments.append(
            ProtectedFragment(
                kind=spec.kind,
                ordinal=ordinal,
                source_span=span,
                source_text=source_map.slice(span),
                marker=marker_text(
                    nonce,
                    spec.kind,
                    ordinal,
                    sql_comment=spec.sql_comment,
                    canonical_newline=spec.canonical_newline,
                ),
                required_offset=spec.required_offset,
            )
        )
        previous_end = span.end
    frozen = tuple(fragments)
    return ProtectionPlan(
        nonce=nonce,
        protected_sql=mask_fragments(
            source_map.slice(literal.content_span),
            literal.content_span.start,
            frozen,
        ),
        fragments=frozen,
    )


_TOKEN_TEMPLATE = r"__INLINE_SQL_{nonce}_[A-Z_]+_[0-9]+__"


def _marker_token(marker: str, nonce: str) -> str:
    match = re.fullmatch(
        _TOKEN_TEMPLATE.format(nonce=re.escape(nonce)), marker.rstrip("\n")
    )
    if marker.startswith("-- "):
        match = re.fullmatch(
            r"-- " + _TOKEN_TEMPLATE.format(nonce=re.escape(nonce)) + r"\n?",
            marker,
        )
    if match is None:
        raise UnsafeRestore("invalid protected marker")
    return match.group(0).removeprefix("-- ").removesuffix("\n")


def _embedded_in_sql(formatted: str, position: int) -> bool:
    """Return whether a token position is inside an SQL quote/comment."""
    quote: str | None = None
    line_comment = False
    block_comment = False
    index = 0
    while index < position:
        character = formatted[index]
        following = formatted[index + 1] if index + 1 < position else ""
        if line_comment:
            if character in "\r\n":
                line_comment = False
            index += 1
            continue
        if block_comment:
            if character == "*" and following == "/":
                block_comment = False
                index += 2
            else:
                index += 1
            continue
        if quote is not None:
            if character == "\\":
                index += 2
                continue
            if character == quote:
                if index + 1 < position and formatted[index + 1] == quote:
                    index += 2
                else:
                    quote = None
            index += 1
            continue
        if character in "'\"`":
            quote = character
            index += 1
        elif character == "-" and following == "-":
            line_comment = True
            index += 2
        elif character == "/" and following == "*":
            block_comment = True
            index += 2
        else:
            index += 1
    return quote is not None or line_comment or block_comment


def restore_protected(formatted: str, plan: ProtectionPlan) -> str:
    """Restore all protected fragments in one validated ordered scan."""
    if not re.fullmatch(r"[0-9a-f]{32}", plan.nonce):
        raise UnsafeRestore("invalid protection nonce")
    if formatted.count(plan.nonce) != len(plan.fragments):
        raise UnsafeRestore("protection namespace count changed")
    namespace = re.compile(_TOKEN_TEMPLATE.format(nonce=re.escape(plan.nonce)))
    actual_tokens = tuple(match.group(0) for match in namespace.finditer(formatted))
    expected_tokens = tuple(
        _marker_token(fragment.marker, plan.nonce) for fragment in plan.fragments
    )
    if actual_tokens != expected_tokens:
        raise UnsafeRestore("protected marker sequence changed")

    pieces: list[str] = []
    cursor = 0
    for fragment, expected_token in zip(plan.fragments, expected_tokens, strict=True):
        expected_identity = (
            f"__INLINE_SQL_{plan.nonce}_{fragment.kind.name}_{fragment.ordinal}__"
        )
        if expected_token != expected_identity:
            raise UnsafeRestore("protected marker identity changed")
        position = formatted.find(fragment.marker, cursor)
        if position < 0:
            raise UnsafeRestore("protected marker spelling changed")
        token_position = position + (3 if fragment.marker.startswith("-- ") else 0)
        if (
            formatted[token_position : token_position + len(expected_token)]
            != expected_token
        ):
            raise UnsafeRestore("protected marker identity changed")
        if fragment.kind is not ProtectedKind.SQL_MARKER and _embedded_in_sql(
            formatted, token_position
        ):
            raise UnsafeRestore("protected marker embedded in SQL")
        if (
            fragment.required_offset is not None
            and position != fragment.required_offset
        ):
            raise UnsafeRestore("anchored marker moved")
        pieces.extend((formatted[cursor:position], fragment.source_text))
        cursor = position + len(fragment.marker)
    pieces.append(formatted[cursor:])
    restored = "".join(pieces)
    if plan.nonce in restored:
        raise UnsafeRestore("protection namespace remains after restore")
    return restored

File: src/python-analysis/protection.ts
import type { SqlDetection } from "./detection.js";
import { SourceMap, SourceSpan } from "./positions.js";
import type { LiteralKind, SupportedLiteral } from "./tokenizer.js";

/** Kind of source text replaced by an opaque formatter marker. */
export type ProtectedKind = "field" | "escaped_brace" | "python_escape" | "sql_marker";

/** A source-free failure to restore protected Python text. */
export class UnsafeRestore extends Error {}

/** One source span replaced by a nonce-bearing marker. */
export interface ProtectedFragment {
  readonly kind: ProtectedKind;
  readonly ordinal: number;
  readonly sourceSpan: SourceSpan;
  readonly sourceText: string;
  readonly marker: string;
  readonly requiredOffset: number | undefined;
}

/** The masked literal body and its exact restoration fragments. */
export interface ProtectionPlan {
  readonly nonce: string;
  readonly protectedSql: string;
  readonly fragments: readonly ProtectedFragment[];
}

/** Allocate a 32-hex-character request nonce absent from *source*. */
export function allocateNonce(source: string, randomHex: () => string): string {
  for (let attempt = 0; attempt < 128; attempt++) {
    const candidate = randomHex();
    if (!/^[0-9a-f]{32}$/.test(candidate)) {
      throw new UnsafeRestore("invalid protection nonce");
    }
    if (!source.includes(candidate)) return candidate;
  }
  throw new UnsafeRestore("unable to allocate protection nonce");
}

/** Return the stable opaque token used for one protected fragment. */
export function markerText(
  nonce: string,
  kind: ProtectedKind,
  ordinal: number,
  options: { readonly sqlComment: boolean; readonly canonicalNewline: boolean },
): string {
  const token = `__INLINE_SQL_${nonce}_${kind.toUpperCase()}_${ordinal}__`;
  if (options.sqlComment) return `-- ${token}${options.canonicalNewline ? "\n" : ""}`;
  return `${token}${options.canonicalNewline ? "\n" : ""}`;
}

function intersectsAny(span: SourceSpan, blocked: readonly SourceSpan[]): boolean {
  return blocked.some((other) => span.start < other.end && other.start < span.end);
}

/** Return the complete source spelling of an escape at *start*. */
function pythonEscapeEnd(source: string, start: number, limit: number): number {
  if (start + 1 >= limit || source[start] !== "\\") {
    throw new UnsafeRestore("invalid Python escape boundary");
  }
  const next = source[start + 1] ?? "";
  if (next === "\r" && start + 2 < limit && source[start + 2] === "\n") return start + 3;
  if (next === "\r" || next === "\n") return start + 2;
  if (next === "N" && start + 2 < limit && source[start + 2] === "{") {
    const close = source.indexOf("}", start + 3);
    if (close < 0 || close >= limit) {
      throw new UnsafeRestore("unterminated named Python escape");
    }
    return close + 1;
  }
  const widths: Record<string, number> = { x: 2, u: 4, U: 8 };
  if (next in widths) {
    const width = widths[next];
    if (width === undefined) {
      throw new UnsafeRestore("unterminated fixed-width Python escape");
    }
    const end = start + 2 + width;
    if (end > limit) throw new UnsafeRestore("unterminated fixed-width Python escape");
    return end;
  }
  if ("01234567".includes(next)) {
    let end = start + 2;
    while (end < Math.min(start + 4, limit) && "01234567".includes(source[end] ?? "")) {
      end++;
    }
    return end;
  }
  return start + 2;
}

interface FragmentSpec {
  readonly kind: ProtectedKind;
  readonly sourceSpan: SourceSpan;
  readonly sqlComment: boolean;
  readonly canonicalNewline: boolean;
  readonly requiredOffset: number | undefined;
}

function spec(
  kind: ProtectedKind,
  sourceSpan: SourceSpan,
  options: Partial<Omit<FragmentSpec, "kind" | "sourceSpan">> = {},
): FragmentSpec {
  return {
    kind,
    sourceSpan,
    sqlComment: options.sqlComment ?? false,
    canonicalNewline: options.canonicalNewline ?? false,
    requiredOffset: options.requiredOffset,
  };
}

/** Discover protected syntax spans without evaluating Python text. */
function discoverSourceSpecs(
  sourceMap: SourceMap,
  literal: SupportedLiteral,
  detection: SqlDetection,
): readonly FragmentSpec[] {
  const content = literal.contentSpan;
  const specs: FragmentSpec[] = [];
  if (detection.markerSpan !== undefined) {
    if (detection.sqlSpan === undefined) {
      throw new UnsafeRestore("marker detection has no SQL boundary");
    }
    const markerSpan = new SourceSpan(content.start, detection.sqlSpan.start);
    const markerSource = sourceMap.slice(markerSpan);
    specs.push(
      spec("sql_marker", markerSpan, {
        sqlComment: true,
        canonicalNewline: /[\r\n]$/.test(markerSource),
        requiredOffset: 0,
      }),
    );
  }
  specs.push(...literal.fieldSpans.map((span) => spec("field", span)));
  const fixedSpans = specs.map((item) => item.sourceSpan);

  if (literal.kind !== "raw" && literal.kind !== "raw_fstring") {
    let cursor = content.start;
    while (cursor < content.end) {
      if (sourceMap.text[cursor] !== "\\") {
        cursor++;
        continue;
      }
      const end = pythonEscapeEnd(sourceMap.text, cursor, content.end);
      const span = new SourceSpan(cursor, end);
      if (!intersectsAny(span, fixedSpans)) {
        specs.push(spec("python_escape", span));
      }
      cursor = end;
    }
  }

  let cursor = content.start;
  while (cursor + 1 < content.end) {
    const pair = sourceMap.text.slice(cursor, cursor + 2);
    if (pair === "{{") {
      const close = sourceMap.text.indexOf("}}", cursor + 2);
      const span = new SourceSpan(cursor, close === -1 ? content.end : close + 2);
      const overlaps = specs.filter((item) => intersectsAny(span, [item.sourceSpan]));
      if (overlaps.length === 0) {
        specs.push(spec("escaped_brace", span));
      } else if (overlaps.every((item) => item.kind === "python_escape")) {
        const merged = new SourceSpan(
          Math.min(span.start, ...overlaps.map((item) => item.sourceSpan.start)),
          Math.max(span.end, ...overlaps.map((item) => item.sourceSpan.end)),
        );
        const overlapSet = new Set(overlaps);
        const filtered = specs.filter((item) => !overlapSet.has(item));
        filtered.push(spec("python_escape", merged));
        specs.length = 0;
        specs.push(...filtered);
      }
      cursor = span.end;
      continue;
    }
    if (pair === "}}") {
      const span = new SourceSpan(cursor, cursor + 2);
      const overlaps = specs.filter((item) => intersectsAny(span, [item.sourceSpan]));
      if (overlaps.length === 0) {
        specs.push(spec("escaped_brace", span));
      }
      cursor += 2;
      continue;
    }
    cursor += 1;
  }
  return [...specs].sort((left, right) => left.sourceSpan.start - right.sourceSpan.start);
}

/** Replace source fragments in order, rejecting overlap or out-of-range spans. */
function maskFragments(
  content: string,
  contentStart: number,
  fragments: readonly ProtectedFragment[],
): string {
  const pieces: string[] = [];
  let cursor = 0;
  for (const fragment of fragments) {
    const start = fragment.sourceSpan.start - contentStart;
    const end = fragment.sourceSpan.end - contentStart;
    if (start < cursor || end < start || end > content.length) {
      throw new UnsafeRestore("protected source spans overlap");
    }
    pieces.push(content.slice(cursor, start), fragment.marker);
    cursor = end;
  }
  pieces.push(content.slice(cursor));
  return pieces.join("");
}

/** Build an exact source mask for one detected SQL literal. */
export function buildProtectionPlan(
  sourceMap: SourceMap,
  literal: SupportedLiteral,
  detection: SqlDetection,
  nonce: string,
): ProtectionPlan {
  if (
    !detection.matched ||
    detection.sqlSpan === undefined ||
    !/^[0-9a-f]{32}$/.test(nonce) ||
    sourceMap.text.includes(nonce)
  ) {
    throw new UnsafeRestore("invalid protection-plan input");
  }
  if (
    literal.contentSpan.start < 0 ||
    literal.contentSpan.end > sourceMap.text.length ||
    detection.sqlSpan.start < literal.contentSpan.start ||
    detection.sqlSpan.end !== literal.contentSpan.end
  ) {
    throw new UnsafeRestore("invalid protection-plan span");
  }
  const specs = discoverSourceSpecs(sourceMap, literal, detection);
  const fragments: ProtectedFragment[] = [];
  let previousEnd = literal.contentSpan.start;
  for (let ordinal = 0; ordinal < specs.length; ordinal++) {
    const item = specs[ordinal];
    if (item === undefined) throw new UnsafeRestore("protected source spans overlap");
    const span = item.sourceSpan;
    if (
      span.start < previousEnd ||
      span.start < literal.contentSpan.start ||
      span.end > literal.contentSpan.end
    ) {
      throw new UnsafeRestore("protected source spans overlap");
    }
    fragments.push({
      kind: item.kind,
      ordinal,
      sourceSpan: span,
      sourceText: sourceMap.slice(span),
      marker: markerText(nonce, item.kind, ordinal, {
        sqlComment: item.sqlComment,
        canonicalNewline: item.canonicalNewline,
      }),
      requiredOffset: item.requiredOffset,
    });
    previousEnd = span.end;
  }
  return {
    nonce,
    protectedSql: maskFragments(
      sourceMap.slice(literal.contentSpan),
      literal.contentSpan.start,
      fragments,
    ),
    fragments,
  };
}

/** Extract the bare token from a marker whose nonce must match *nonce*. */
function markerToken(marker: string, nonce: string): string {
  const escapedNonce = nonce.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const tokenPattern = `__INLINE_SQL_${escapedNonce}_[A-Z_]+_[0-9]+__`;
  if (marker.startsWith("-- ")) {
    const match = new RegExp(`^-- ${tokenPattern}\\n?$`).exec(marker);
    if (match === null) throw new UnsafeRestore("invalid protected marker");
    return match[0].replace(/^-- /, "").replace(/\n$/, "");
  }
  const match = new RegExp(`^${tokenPattern}\\n?$`).exec(marker);
  if (match === null) throw new UnsafeRestore("invalid protected marker");
  return match[0].replace(/\n$/, "");
}

/** Restore all protected fragments in one validated ordered scan. */
export function restoreProtected(formatted: string, plan: ProtectionPlan): string {
  if (!/^[0-9a-f]{32}$/.test(plan.nonce)) {
    throw new UnsafeRestore("invalid protection nonce");
  }
  if (formatted.split(plan.nonce).length - 1 !== plan.fragments.length) {
    throw new UnsafeRestore("protection namespace count changed");
  }
  const escapedNonce = plan.nonce.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const namespace = new RegExp(`__INLINE_SQL_${escapedNonce}_[A-Z_]+_[0-9]+__`, "g");
  const actualTokens = [...formatted.matchAll(namespace)].map((match) => match[0]);
  const expectedTokens = plan.fragments.map((fragment) => markerToken(fragment.marker, plan.nonce));
  if (actualTokens.join("\u0000") !== expectedTokens.join("\u0000")) {
    throw new UnsafeRestore("protected marker sequence changed");
  }

  const pieces: string[] = [];
  let cursor = 0;
  plan.fragments.forEach((fragment, index) => {
    const expectedToken = expectedTokens[index];
    if (expectedToken === undefined) {
      throw new UnsafeRestore("protected marker sequence changed");
    }
    const expectedIdentity = `__INLINE_SQL_${plan.nonce}_${fragment.kind.toUpperCase()}_${fragment.ordinal}__`;
    if (expectedToken !== expectedIdentity) {
      throw new UnsafeRestore("protected marker identity changed");
    }
    const position = formatted.indexOf(fragment.marker, cursor);
    if (position < 0) throw new UnsafeRestore("protected marker spelling changed");
    let tokenPosition = position;
    if (fragment.marker.startsWith("-- ")) tokenPosition += 3;
    else if (fragment.marker.startsWith('"')) tokenPosition += 1;
    if (formatted.slice(tokenPosition, tokenPosition + expectedToken.length) !== expectedToken) {
      throw new UnsafeRestore("protected marker identity changed");
    }
    if (fragment.requiredOffset !== undefined && position !== fragment.requiredOffset) {
      throw new UnsafeRestore("anchored marker moved");
    }
    pieces.push(formatted.slice(cursor, position), fragment.sourceText);
    cursor = position + fragment.marker.length;
  });
  pieces.push(formatted.slice(cursor));
  const restored = pieces.join("");
  if (restored.includes(plan.nonce)) {
    throw new UnsafeRestore("protection namespace remains after restore");
  }
  return restored;
}

export type { LiteralKind };
